src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -283,9 +285,17 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                     }
                     else
                     {
-                        // Send EndStream asynchronously and without cancellation.
-                        // If this fails, it means that the connection is aborting and we will be reset.
-                        _connection.LogExceptions(_connection.SendEndStreamAsync(StreamId));
+                        // Use dedicated lock to prevent sending a frame with EndStream flag after an RST_STREAM frame.
+                        lock (_eofSyncObj)
```

Comment:
I don't think we need a separate lock here.

With the changes we made to write queuing in 6.0, the call below to SendEndStreamAsync is simply going to enqueue the write operation in the write queue. Similarly for SendResetStreamAsync. As such, I think it's fine to have these happen under the main SyncObject lock.

Reply:
Initially, I thought about reusing SyncObj for locking, but then I saw that there is [Debug.Assert in SendReset](https://github.com/dotnet/runtime/blob/183c4d100f68fb6c177a1fe71809d581aa25e47b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L328) and also I found your PR dotnet/corefx#40123 where you added that assert explaining the reason for sending RST_STREAM and EndStream outside lock.
Could you please clarify if that concern doesn't apply anymore?

Reply:
The issue in https://github.com/dotnet/corefx/pull/40123 doesn't apply anymore because we never do the write on the calling thread; we just add an entry to the write queue.

Reply:
Ok. Got it. I will change it to use lock(SyncObj), then.

Reply:
@geoffkizer I made the requested change. Please, review.
There is a few failed tests, but they are unrelated to this change.